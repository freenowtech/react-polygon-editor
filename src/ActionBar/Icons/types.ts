import { SVGProps } from 'react';

export interface IconProps extends SVGProps<SVGSVGElement> {
    iconColor?: string;
}
