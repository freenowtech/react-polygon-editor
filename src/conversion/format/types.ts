import { Coordinate } from 'types';

export enum FormatType {
    GEOJSON = 'geojson',
    LATLNG = 'latlng',
    JTS = 'jts',
}

export interface Format {
    name: FormatType;
    displayName: string;
    description?: string;
    serialize: (coordinates: Coordinate[]) => string;
    deserialize: (raw: string) => Coordinate[];
    validate: (value: string) => boolean;
}
