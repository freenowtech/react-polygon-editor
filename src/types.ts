import { ActionCreatorsMapObject } from 'redux';

export type Coordinate = {
    latitude: number;
    longitude: number;
};

export interface Action<T extends string> {
    type: T;
}

export interface ActionWithPayload<T extends string, P> extends Action<T> {
    payload: P;
}

export type ActionsUnion<A extends ActionCreatorsMapObject> = ReturnType<A[keyof A]>;
