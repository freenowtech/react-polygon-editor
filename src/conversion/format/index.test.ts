import path from 'path';
import fs from 'fs';

import { FormatType } from './types';
import { format } from '.';

const testdataDir = path.resolve(__dirname, '__testdata__');

type TestCase = [
    string, // name
    'valid' | 'invalid', // state
    string, // path
    FormatType // format
];

describe('conversion format', () => {
    const formats = Object.keys(format);

    const getTestCaseFromFile = (filePath: string): TestCase => {
        const fileName = path.basename(filePath, '.json');
        const [state, formatType, testName] = fileName.split('_');

        if (state !== 'valid' && state !== 'invalid') {
            throw new Error(`test file ${fileName}: must begin with 'valid_' or 'invalid_'`);
        }

        if (formats.every((f) => f !== formatType)) {
            throw new Error(`test file ${fileName}: format must be in ${formats}, but was ${formatType}`);
        }

        if (!testName) {
            throw new Error(`test file ${fileName}: format must have a test name`);
        }

        return [testName.replace('-', ' '), state, filePath, formatType as FormatType];
    };

    const getTestCasesForFormat = (f: FormatType): TestCase[] => {
        return fs
            .readdirSync(testdataDir)
            .map(getTestCaseFromFile)
            .filter((testCase) => testCase[3] === f);
    };

    describe.each(formats)('%s', (f: FormatType) => {
        const testCases = getTestCasesForFormat(f);

        if (testCases.length > 0) {
            test.each(testCases)('validation: %s => %s', (testName, state, testPath, formatType) => {
                const raw = fs.readFileSync(path.join(testdataDir, testPath)).toString('utf-8');
                const expected = state === 'valid';
                const actual = format[formatType].validate(raw);
                expect(actual).toBe(expected);
            });

            test.each(testCases)('deserialize: %s => %s', (testName, state, testPath, formatType) => {
                const raw = fs.readFileSync(path.join(testdataDir, testPath)).toString('utf-8');
                if (state === 'valid') {
                    expect(format[formatType].deserialize(raw)).toMatchSnapshot();
                } else {
                    expect(() => format[formatType].deserialize(raw)).toThrowErrorMatchingSnapshot();
                }
            });
        }
    });
});
