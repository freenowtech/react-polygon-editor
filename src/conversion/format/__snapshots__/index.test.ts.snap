// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conversion format geojson deserialize: empty file => invalid 1`] = `"Unexpected end of JSON input"`;

exports[`conversion format geojson deserialize: missing coordinates => invalid 1`] = `"Cannot read properties of undefined (reading 'length')"`;

exports[`conversion format geojson deserialize: missing type => invalid 1`] = `
"Invalid GeoJSON detected:
- "type" member required
"
`;

exports[`conversion format geojson deserialize: no right-hand-rule-should-still-pass => valid 1`] = `
[
  {
    "latitude": 0,
    "longitude": 0,
  },
  {
    "latitude": 1,
    "longitude": 1,
  },
  {
    "latitude": 1,
    "longitude": 0,
  },
  {
    "latitude": 0,
    "longitude": 0,
  },
]
`;

exports[`conversion format geojson deserialize: not enough-coordinates => invalid 1`] = `
"Invalid GeoJSON detected:
- a LinearRing of coordinates needs to have four or more positions
,- Polygons and MultiPolygons should follow the right-hand rule
"
`;

exports[`conversion format geojson deserialize: open polygon => invalid 1`] = `
"Invalid GeoJSON detected:
- the first and last positions in a LinearRing of coordinates must be the same
"
`;

exports[`conversion format geojson deserialize: type feature => valid 1`] = `
[
  {
    "latitude": 53.56545917520555,
    "longitude": 10.0412155687809,
  },
  {
    "latitude": 53.56586697476003,
    "longitude": 10.018899589776995,
  },
  {
    "latitude": 53.58380626282291,
    "longitude": 10.005166679620745,
  },
  {
    "latitude": 53.56668256207333,
    "longitude": 9.993150383234026,
  },
  {
    "latitude": 53.56647866671947,
    "longitude": 9.961908012628557,
  },
  {
    "latitude": 53.555466857846014,
    "longitude": 9.979417473077776,
  },
  {
    "latitude": 53.55505895802228,
    "longitude": 10.02576604485512,
  },
  {
    "latitude": 53.56545917520555,
    "longitude": 10.0412155687809,
  },
]
`;

exports[`conversion format geojson deserialize: type feature-collection => valid 1`] = `
[
  {
    "latitude": 53.56545917520555,
    "longitude": 10.0412155687809,
  },
  {
    "latitude": 53.56586697476003,
    "longitude": 10.018899589776995,
  },
  {
    "latitude": 53.58380626282291,
    "longitude": 10.005166679620745,
  },
  {
    "latitude": 53.56668256207333,
    "longitude": 9.993150383234026,
  },
  {
    "latitude": 53.56647866671947,
    "longitude": 9.961908012628557,
  },
  {
    "latitude": 53.555466857846014,
    "longitude": 9.979417473077776,
  },
  {
    "latitude": 53.55505895802228,
    "longitude": 10.02576604485512,
  },
  {
    "latitude": 53.56545917520555,
    "longitude": 10.0412155687809,
  },
]
`;

exports[`conversion format geojson deserialize: type polygon => valid 1`] = `
[
  {
    "latitude": 53.56545917520555,
    "longitude": 10.0412155687809,
  },
  {
    "latitude": 53.56586697476003,
    "longitude": 10.018899589776995,
  },
  {
    "latitude": 53.58380626282291,
    "longitude": 10.005166679620745,
  },
  {
    "latitude": 53.56668256207333,
    "longitude": 9.993150383234026,
  },
  {
    "latitude": 53.56647866671947,
    "longitude": 9.961908012628557,
  },
  {
    "latitude": 53.555466857846014,
    "longitude": 9.979417473077776,
  },
  {
    "latitude": 53.55505895802228,
    "longitude": 10.02576604485512,
  },
  {
    "latitude": 53.56545917520555,
    "longitude": 10.0412155687809,
  },
]
`;
